lib/SILOptimizer/Utils/Existential.cpp

```diff
@@ -311,3 +312,83 @@ ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &openedUse) {
+                                                 SILBuilder &Builder)
+    : ConcreteType(ConcreteTy) {
+  SILValue Arg = ArgOperand.get();
+  SILInstruction *User = ArgOperand.getUser();
+
+  // We need the OpenedArchetypeDef to be able to cast it to the concrete type.
+  // findInitExistential helps us determine this OpenedArchetypeDef. For cases
+  // where OpenedArchetypeDef can not be found from findInitExistential (because
+  // there was no InitExistential), then the code after findInitExistential does
+  // extra effort in trying to find an OpenedArchetypeDef for AllocStackInst
+  // using getAddressOfStackInit.
+  isCopied = false;
+  OpenedArchetypeDef = Arg;
+  InitExistential = findInitExistential(ArgOperand, OpenedArchetype,
+                                        OpenedArchetypeDef, isCopied);
+
+  // If findInitExistential did not update the OpenedArchetypeDef (because it
+  // did not find an InitExistential) and that the original Arg is an
+  // alloc_stack instruction, then we determine the OpenedArchetypeDef using
+  // getAddressOfStackInit. Please keep in mind that an alloc_stack can be
+  // an argument to apply (wich has no InitExistential), thus we need to
+  // determine the OpenedArchetypeDef for it. This is the extra effort.
+  if (auto *Instance = dyn_cast<AllocStackInst>(OpenedArchetypeDef)) {
+    bool copied = false;
+    if (SILValue Src = getAddressOfStackInit(Instance, User, copied)) {
+      OpenedArchetypeDef = Src;
+    }
+    isCopied = copied;
+  }
+
+  // Bail, if we did not find an opened existential.
+  if (!(isa<OpenExistentialRefInst>(OpenedArchetypeDef) ||
+        isa<OpenExistentialAddrInst>(OpenedArchetypeDef)))
+    return;
+
+  // Create SIL type for the concrete type.
+  auto &M = Builder.getModule();
+  SILType ConcreteSILType = M.Types.getLoweredType(ConcreteType);
+
+  // We have the open_existential; we still need the conformance.
+  auto ConformanceRef =
+      M.getSwiftModule()->lookupConformance(ConcreteType, Protocol);
+  if (!ConformanceRef)
+    return;
+
+  /// Determine the ExistentialConformances and SubstitutionMap.
+  auto ExistentialConformances = ArrayRef<ProtocolConformanceRef>(
+      ProtocolConformanceRef(ConformanceRef.getValue().getConcrete()));
+  ExistentialType = Protocol->getDeclaredType()->getCanonicalType();
+  auto ExistentialSig = M.getASTContext().getExistentialSignature(
+      ExistentialType, M.getSwiftModule());
+  ExistentialSubs = SubstitutionMap::get(ExistentialSig, {ConcreteType},
+                                         ExistentialConformances);
+
+  // Prepare the code by adding UncheckedCast instructions that cast opened
+  // existentials to concrete types.
+  if (auto *OER = dyn_cast<OpenExistentialRefInst>(OpenedArchetypeDef)) {
+    auto *URCI =
+        Builder.createUncheckedRefCast(OER->getLoc(), OER, ConcreteSILType);
+    OpenedArchetype = OER->getType().getASTType()->castTo<ArchetypeType>();
+    ConcreteValue = URCI;
```

Comment:
Ok, this works, but it's kind of hacky because ConcreteExistentialInfo is supposed to be a mini-analysis that just summarizes all the various SIL patterns we may encounter where the common thread is that we have an open_existential as the `self` argument of an apply. It's confusing if an analysis utility also mutates the SIL. Rather than passing in a Builder, `propagateSoleConformingType` should be responsible for generating the cast that it needs. I think it's fine if you allow a valid ConcreteExistentialInfo to have an invalid ConcreteValue when InitExistential == nullptr.

So maybe you could remove `ConcreteValue` from the `isValid` and add an assert instead: `!InitExistential || ConcreteValue`

Then, in your new CHA/PCA case, you could use just use the existing ConcreteValue if an InitExistential is found without doing the cast. Would that make sense? If there's no InitExistential, then `propagateSoleConformingType` could generate the cast.